amethyst_renderer/src/pass/flat2d/interleaved.rs

```diff
@@ -0,0 +1,506 @@
+        flipped: Option<&Flipped>,
+        tex_storage: &AssetStorage<Texture>,
+    ) {
+        let global = match global {
+            Some(v) => v,
+            None => return,
+        };
+
+        let texture_dims = match tex_storage.get(&texture_handle) {
+            Some(tex) => tex.size(),
+            None => {
+                warn!("Texture not loaded for texture: `{:?}`.", texture_handle);
+                return;
+            }
+        };
+
+        self.textures.push(TextureDrawData::Image {
+            texture_handle: texture_handle.clone(),
+            transform: *global,
+            flipped: flipped.map(|it| it.clone()),
+            width: texture_dims.0,
+            height: texture_dims.1,
+        });
+    }
+
+    pub fn add_sprite(
+        &mut self,
+        sprite_render: &SpriteRender,
+        global: Option<&GlobalTransform>,
+        flipped: Option<&Flipped>,
+        sprite_sheet_storage: &AssetStorage<SpriteSheet>,
+        tex_storage: &AssetStorage<Texture>,
+    ) {
+        let global = match global {
+            Some(v) => v,
+            None => return,
+        };
+
+        let texture_handle = match sprite_sheet_storage.get(&sprite_render.sprite_sheet) {
+            Some(sprite_sheet) => {
+                if tex_storage.get(&sprite_sheet.texture).is_none() {
+                    warn!(
+                        "Texture not loaded for texture: `{:?}`.",
+                        sprite_sheet.texture
+                    );
+                    return;
+                }
+
+                sprite_sheet.texture.clone()
+            }
+            None => {
+                warn!(
+                    "Sprite sheet not loaded for sprite_render: `{:?}`.",
+                    sprite_render
+                );
+                return;
+            }
+        };
+
+        self.textures.push(TextureDrawData::Sprite {
+            texture_handle,
+            render: sprite_render.clone(),
+            flipped: flipped.map(|it| it.clone()),
+            transform: *global,
+        });
+    }
+
+    /// Optimize the sprite order to generating more coherent batches.
+    pub fn sort(&mut self) {
+        // Only takes the texture into account for now.
+        self.textures.sort_by(|a, b| a.tex_id().cmp(&b.tex_id()));
+    }
+
+    pub fn encode(
+        &self,
+        encoder: &mut Encoder,
+        factory: &mut Factory,
+        effect: &mut Effect,
+        camera: Option<(&Camera, &GlobalTransform)>,
+        sprite_sheet_storage: &AssetStorage<SpriteSheet>,
+        tex_storage: &AssetStorage<Texture>,
+    ) {
+        use gfx::{
+            buffer,
+            memory::{Bind, Typed},
+            Factory,
+        };
+
+        if self.textures.is_empty() {
+            return;
+        }
+
+        // Sprite vertex shader
+        set_view_args(effect, encoder, camera);
+
+        let mut instance_data = Vec::<f32>::new();
+        let mut num_instances = 0;
```

Comment:
Actually `num_instances == instance_data.len() / 11`, because each time we add instance data we add 11 `f32`s. I think this is easier to read and results in the same performance in the end, probably.

Reply:
Oh I see, you're right about read-ability